File: napa/TripsViewController.m
//
//  MasterViewController.m
//  napa
//
//  Created by Remy JOURDE on 25/05/2014.
//  Copyright (c) 2014 Remy Jourde. All rights reserved.
//

#import "TripsViewController.h"

#import "TripsViewControllerDataSource.h"
#import "NewEditTripViewController.h"
#import "DetailViewController.h"
#import "Trip.h"
#import "TripViewCell.h"

static NSString *kShowTripSegue     = @"showTrip";
static NSString *kNewTripSegue      = @"newTrip";
static NSString *kEditTripSegue     = @"editTrip";

#define kDeleteButtonIndex      0
#define kEditButtonIndex        1
#define kCancelButtonIndex      2

@interface TripsViewController () <TripsViewControllerDataSourceDelegate, NewEditTripViewControllerDelegate>

@property (nonatomic, strong) TripsViewControllerDataSource* tripsViewControllerDataSource;

// keep track of the selected collection view cell
@property (nonatomic, strong) NSIndexPath *selectedIndexPath;

@property (nonatomic, strong) Trip *currentTrip;

- (void)editTrip;
- (void)deleteTrip;

@end

@implementation TripsViewController


- (void)awakeFromNib
{
    [super awakeFromNib];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // attach long press gesture to collectionView
    UILongPressGestureRecognizer *lpgr = [[UILongPressGestureRecognizer alloc]
                                          initWithTarget:self action:@selector(handleLongPress:)];
    lpgr.minimumPressDuration = .5; //seconds
    lpgr.delegate = self;
    [self.collectionView addGestureRecognizer:lpgr];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [self setupFetchedResultsController];
}

- (void)handleLongPress:(UILongPressGestureRecognizer *)gestureRecognizer
{
    if (gestureRecognizer.state == UIGestureRecognizerStateBegan)
    {
        CGPoint point = [gestureRecognizer locationInView:self.collectionView];
        
        self.selectedIndexPath = [self.collectionView indexPathForItemAtPoint:point];
        if (self.selectedIndexPath == nil)
        {
            NSLog(@"couldn't find index path");
        }
        else
        {
            UIActionSheet *actionSheet = [[UIActionSheet alloc] initWithTitle:nil delegate:self
                                                            cancelButtonTitle:@"Cancel"
                                                       destructiveButtonTitle:@"Delete Trip"
                                                            otherButtonTitles:@"Edit Trip", nil];
            [actionSheet showInView:self.view];
        }
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setupFetchedResultsController
{
    self.tripsViewControllerDataSource = [[TripsViewControllerDataSource alloc] initWithCollectionView:self.collectionView];
    self.tripsViewControllerDataSource.fetchedResultsController = [Trip tripsFetchedResultsControllerInManagedObjectContext:self.managedObjectContext];
    self.tripsViewControllerDataSource.delegate = self;
}

#pragma mark - Fetched Results Controller Delegate

- (void)configureCell:(id)theCell withObject:(id)object
{
    TripViewCell* cell = theCell;
    Trip* trip = object;
    cell.nameLabel.text = trip.name;
    
    NSInteger days = [trip.startDate numberOfDaysUntilDay:trip.endDate];
    
    if (days <= 1) {
        cell.daysLabel.text = [NSString stringWithFormat:@"1 day"];
    }
    else
    {
        cell.daysLabel.text = [NSString stringWithFormat:@"%ld days", days];
    }
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"dd"];
    cell.dateLabel.text = [dateFormatter stringFromDate:trip.startDate];
    
    [dateFormatter setDateFormat:@"MMM"];
    cell.monthLabel.text = [dateFormatter stringFromDate:trip.startDate];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    [super prepareForSegue:segue sender:sender];
    if ([segue.identifier isEqualToString:kShowTripSegue])
    {
        if (!self.currentTrip)
        {
            [self presentTripViewController:segue.destinationViewController];
        }
        else
        {
            [(DetailViewController*)segue.destinationViewController setDetailItem:self.currentTrip];
            self.currentTrip = nil;
        }
    }
    if ([segue.identifier isEqualToString:kNewTripSegue])
    {
        UINavigationController *navigationController = segue.destinationViewController;
        NewEditTripViewController *newTripViewController = [[navigationController viewControllers] objectAtIndex:0];
        newTripViewController.managedObjectContext = self.managedObjectContext;
        newTripViewController.delegate = self;
    }
    if ([segue.identifier isEqualToString:kEditTripSegue])
    {
        UINavigationController *navigationController = segue.destinationViewController;
        NewEditTripViewController *editTripViewController = [[navigationController viewControllers] objectAtIndex:0];
        editTripViewController.managedObjectContext = self.managedObjectContext;
        editTripViewController.editModeON = YES;
        editTripViewController.objectToEdit = [self.tripsViewControllerDataSource itemAtIndexPath:self.selectedIndexPath];
        editTripViewController.delegate = self;
    }
}

- (void)presentTripViewController:(DetailViewController*)tripViewController
{
    NSManagedObject *object = [self.tripsViewControllerDataSource selectedItem];
    [tripViewController setDetailItem:object];
}

#pragma mark - NewTripViewController delegate

- (void)newTripViewControllerDidCancel:(NewEditTripViewController *)controller
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)newTripViewController:(NewEditTripViewController *)controller didDoneWithTrip:(Trip*) trip
{
    [self dismissViewControllerAnimated:YES completion:nil];
    
    self.currentTrip = trip;
    [self performSegueWithIdentifier:kShowTripSegue sender:self];
}

#pragma mark - UIActionSheet deletegate

- (void)actionSheet:(UIActionSheet *)actionSheet clickedButtonAtIndex:(NSInteger)buttonIndex
{
    switch (buttonIndex) {
        case kDeleteButtonIndex:
            [self deleteTrip];
            break;
        case kEditButtonIndex:
            [self editTrip];
        default:
            break;
    }
}

- (void)editTrip
{
    [self performSegueWithIdentifier:kEditTripSegue sender:self];
}

- (void)deleteTrip
{
	Trip *trip = [self.tripsViewControllerDataSource itemAtIndexPath:self.selectedIndexPath];
    NSString* actionName = [NSString stringWithFormat:@"%@ trip deletion", trip.name];
    [self.undoManager setActionName:actionName];
    [self.managedObjectContext deleteObject:trip];
}

#pragma mark - Undo

- (BOOL)canBecomeFirstResponder
{
    return YES;
}

- (NSUndoManager*)undoManager
{
    return self.managedObjectContext.undoManager;
}

@end

File: napa/TripsViewController.swift
//
//  TripsViewController.swift
//  napa
//
//  Created by Remy JOURDE on 01/10/2014.
//  Copyright (c) 2014 Remy Jourde. All rights reserved.
//

import UIKit
import CoreData

class TripsViewController: UICollectionViewController, UIGestureRecognizerDelegate, UIActionSheetDelegate, TripsViewControllerDataSourceDelegate, NewEditTripViewControllerDelegate {
    
    enum SegueIdentifier: String {
        case ShowTripSegueIdentifier    = "showTrip"
        case NewTripSegueIdentifier     = "newTrip"
        case EditTripSegueIdentifier    = "editTrip"
    }
    
    enum ButtonIndex: Int {
        case DeleteButtonIndex  = 0
        case EditButtonIndex    = 1
        case CancelButtonIndex  = 2
    }
    
    var managedObjectContext: NSManagedObjectContext?
    var tripsViewControllerDataSource: TripsViewControllerDataSource?
    
    // keep track of the selected collection view cell
    var selectedIndexPath: NSIndexPath?
    
    var currentTrip: Trip?

    override func viewDidLoad() {
        super.viewDidLoad()

        // attach long press gesture to collectionView
        let lpgr = UILongPressGestureRecognizer(target: self, action: "handleLongPress")
        lpgr.minimumPressDuration = 0.5 //seconds
        lpgr.delegate = self
        self.collectionView?.addGestureRecognizer(lpgr)
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        self.setupFetchedResultsController()
    }
    
    func handleLongPress(gestureRecognizer: UILongPressGestureRecognizer) {
        if gestureRecognizer.state == UIGestureRecognizerState.Began {
            let point = gestureRecognizer.locationInView(self.collectionView)
    
            self.selectedIndexPath = self.collectionView?.indexPathForItemAtPoint(point)
            if (self.selectedIndexPath != nil)
            {
                let message = UIAlertController(title: nil, message: nil, preferredStyle: UIAlertControllerStyle.ActionSheet)
                message.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: nil))
                message.addAction(UIAlertAction(title: "Delete Trip", style: .Destructive){ (action) in self.deleteTrip() })
                message.addAction(UIAlertAction(title: "Edit Trip", style: .Default){ (action) in self.editTrip() })
                
                presentViewController(message, animated: true, completion: nil)
            }
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func setupFetchedResultsController() {
        self.tripsViewControllerDataSource = TripsViewControllerDataSource(collectionView: self.collectionView!)
        self.tripsViewControllerDataSource?.fetchedResultsController = Trip.tripsFetchedResultsControllerInManagedObjectContext(self.managedObjectContext!)
        self.tripsViewControllerDataSource?.delegate = self
    }
    
    // MARK: - Fetched Results Controller Delegate
    
    func configureCell(cell: AnyObject, object: AnyObject)
    {
        let tripViewCell: TripViewCell = cell as TripViewCell
        let trip: Trip = object as Trip
        tripViewCell.nameLabel?.text = trip.name
    
        let days = trip.endDate - trip.startDate
        if days <= 1 {
            tripViewCell.daysLabel?.text = "1 day"
        } else {
            tripViewCell.daysLabel?.text = "\(days) days"
        }
    
        let dateFormatter = NSDateFormatter()
        dateFormatter.dateFormat = "dd"
        tripViewCell.dateLabel?.text = dateFormatter.stringFromDate(trip.startDate)
    
        dateFormatter.dateFormat = "MMM"
        tripViewCell.monthLabel?.text = dateFormatter.stringFromDate(trip.startDate)
    }

    // MARK: - Navigation
    
    func presentTripViewController(tripViewController: TripViewController) {
        let trip = self.tripsViewControllerDataSource?.selectedItem() as Trip
        tripViewController.setTrip(trip)
    }

    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject!) {
        super.prepareForSegue(segue, sender:sender)
        if let segueIdentifier =  SegueIdentifier.fromRaw(segue.identifier) {
            switch segueIdentifier {
            case .ShowTripSegueIdentifier:
                if let currentTrip = self.currentTrip {
                    let tripViewController = segue.destinationViewController as TripViewController
                    tripViewController.setTrip(currentTrip)
                } else {
                    self.presentTripViewController(segue.destinationViewController as TripViewController)
                }
            case .NewTripSegueIdentifier:
                let navigationController = segue.destinationViewController as UINavigationController
                let newTripViewController = navigationController.viewControllers[0] as NewEditTripViewController
                newTripViewController.managedObjectContext = self.managedObjectContext!
                newTripViewController.delegate = self
            case .EditTripSegueIdentifier:
                let navigationController = segue.destinationViewController as UINavigationController
                let editTripViewController = navigationController.viewControllers[0] as NewEditTripViewController
                editTripViewController.managedObjectContext = self.managedObjectContext!
                if let selectedIndexPath = self.selectedIndexPath? {
                    editTripViewController.trip = self.tripsViewControllerDataSource?.itemAtIndexPath(selectedIndexPath) as Trip?
                }
                editTripViewController.delegate = self
            }
        }
    }
    
    // MARK: - NewTripViewController delegate
    
    func newTripViewControllerDidCancel(controller: NewEditTripViewController) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }
    
    func newTripViewController(controller: NewEditTripViewController, didDoneWith trip: Trip) {
        self.dismissViewControllerAnimated(true, completion:nil)
    
        self.currentTrip = trip
        self.performSegueWithIdentifier(SegueIdentifier.ShowTripSegueIdentifier.toRaw(), sender:self)
    }
    
    // MARK: - UIAlertController action
    
    func editTrip() {
        self.performSegueWithIdentifier(SegueIdentifier.EditTripSegueIdentifier.toRaw(), sender:self)
    }
    
    func deleteTrip() {
        if let selectedIndexPath = self.selectedIndexPath {
            let trip = self.tripsViewControllerDataSource?.itemAtIndexPath(selectedIndexPath) as Trip
            self.managedObjectContext?.undoManager?.setActionName("\(trip.name) trip deletion")
            self.managedObjectContext?.deleteObject(trip)
        }
    }
    
    // MARK: - Undo
    
    override func canBecomeFirstResponder() -> Bool {
        return true
    }
}
